go/pkg/cas/upload.go

```diff
@@ -137,22 +192,30 @@ func (r *UploadResult) Digest(ps *PathSpec) (*digest.Digest, error) {
 		return nil, errors.Errorf("%q is not absolute", ps.Path)
 	}
 
+	//lint:ignore ST1008 this is consistent with cache.SingleFlight.Load() signature.
+	loadCache := func(key interface{}) (*digest.Digest, error, bool) {
+		val, err, loaded := r.u.fsCache.Load(key)
+		var digPtr *digest.Digest
+		if val != nil {
+			dig := digest.NewFromProtoUnvalidated(val.(*digested).digest)
+			digPtr = &dig
+		}
+		return digPtr, err, loaded
+	}
+
+	// Try the ps-as-pointer first.
+	if dig, err, loaded := loadCache(ps); loaded {
+		return dig, err
+	}
+
 	// TODO(nodir): cache this syscall too.
 	info, err := os.Lstat(ps.Path)
 	if err != nil {
 		return nil, err
 	}
 
-	key := makeFSCacheKey(ps.Path, info.Mode().IsRegular(), ps.Exclude)
-	switch val, err, loaded := r.u.fsCache.Load(key); {
-	case !loaded:
-		return nil, nil
-	case err != nil:
-		return nil, err
-	default:
-		dig := digest.NewFromProtoUnvalidated(val.(*digested).digest)
-		return &dig, nil
-	}
+	dig, err, _ := loadCache(makeFSCacheKey(ps.Path, info.Mode().IsRegular(), ps.Exclude))
+	return dig, err
 }
 
 // Upload uploads all files/directories specified by pathC.
```

Comment:
Regarding `Upload` - do we want to return a channel instead of an `*UploadResult`? Otherwise, I think we lose the benefits of having a pathC channel input. The user likely wants to get the results of its uploads (including dirs using allowlist) before it is done "uploading all things".

Reply:
I think we discussed this before. Earlier I suggested `Digest` callback in `UploadInput` for exactly this reason.
I think we agreed that we can add it on demand.

I think a channel return value is a bit hard to work with:
- even if you don't care about having digests in a streaming fashion, you are forced to setup a separate goroutine that receives from the channel. IMO in simple cases the code should be simple too.
- suppose PathSpecs are generated by different goroutines and send to pathC. Even if we channel returned by Upload(), then can't easily know the digest of the root. The programmer would have to somehow figure out which goroutine is interested in a particular digest received from the channel.

That's why I initially proposed `UploadItem.Digest` callback. It covers both concerns:
- If you don't care, then you just don't initialize it. Life is simple.
- Each goroutine may pass their callback.

So I agree that currently we are not benefiting from the fact that pathC is a channel, but I think the way to solve it is to add callback to PathSpec.

Thoughts?

Reply:
see the other thread. 
Impl in https://github.com/bazelbuild/remote-apis-sdks/pull/363 seems pretty simple to me. Storing callbacks to call later would be more complicated.
Also a channel for "digests are ready to be read" seems easier to use (see the other thread).

Reply:
now that each `UploadInput` has `DigestsComputed`, the underlying problem is solved.